crates/re_arrow_store/src/store_gc.rs

```diff
@@ -125,38 +198,258 @@ impl DataStore {
+                            .iter()
+                            .enumerate()
+                            .rev()
+                            .filter_map(|(row_index, cell)| {
+                                cell.as_ref().and_then(|cell| {
+                                    table.col_row_id.get(row_index).map(|row| (row, cell))
+                                })
+                            });
+
+                    for (row_id, cell) in component_rows.by_ref().take(*count) {
+                        *count -= 1;
+
+                        candidate_protected.insert(*row_id);
+
+                        // If the last cell we would otherwise protect is empty, this is something that can potentially
+                        // be removed from the store altogether. This allows us to fully drop tables in cases such as
+                        // removing space-views that exceed the undo threshold.
+                        //
+                        // This is a situation that shows up when completely removing entities, such as deleting
+                        // a space-view. There's no value in holding onto the empty tombstone once it's not masking
+                        // any real data, and repeatedly resetting a blueprint can potentially leave many orphaned
+                        // values since every recreated space-view gets a unique random id.
+                        //
+                        // There is an edge-case to be careful about where a multi-component row is protected by an
+                        // older component.  Even if the newer component was cleared with no further history, the
+                        // protection of the older component will keep it from being cleared, and if we don't protect
+                        // the tombstone then the component will effectively be resurrected.
+                        if *count == 0 && cell.is_empty() {
+                            // Collect all the remaining rows for this component in the masked set. If any of these is
+                            // protected, then we can't actually clear this component.
+                            masked_rows_from_clear
+                                .insert(*row_id, component_rows.map(|(r, _)| *r).collect());
+                            // This only happens once when count is 0. Break here due to iterator lifetime.
+                            break;
+                        }
+                    }
+                }
+            }
+
+            // Now that we've processed every component for this table, we can check if any of the masked
+            // rows are protected. If not, we are free to drop protection from this row as well and allow
+            // the table to be potentially GC'd.
+            for (row, masked_rows) in masked_rows_from_clear {
+                if candidate_protected.is_disjoint(&masked_rows) {
+                    candidate_protected.remove(&row);
+                }
+            }
```

Comment:
I don't follow any of this, despite all the comments 😭 

> If the last cell we would otherwise protect is empty, this is something that can potentially
> be removed from the store altogether. This allows us to fully drop tables in cases such as
> removing space-views that exceed the undo threshold.

Detecting such cases of "dead" entities sounds like it could be the job of a separate function (run after the GC), in order to keep this function simpler?


> There is an edge-case to be careful about where a multi-component row is protected by an older component.

What do you mean by "protected" here? What does it mean for one component to be protected by another older one? Aren't all components in a multi-component row of the same age?

Are you talking about two different types of protections at the same time? Would it be more clear if this function was renamed from `find_all_protected_rows` to `find_latest_n_rows_per_component` or something?

Reply:
> Detecting such cases of "dead" entities sounds like it could be the job of a separate function

Great suggestion. This should simplify the code a lot.

> What does it mean for one component to be protected by another older one?

Consider the 2 cases:
```
row | C1   |  C2
------------------
1   |  X   |  null
2   | null |  Y
3   | null | CLEAR
```
Here, we can successfully GC all the way down to:
```
row | C1   |  C2
------------------
1   |  X   |  null
```

However here,
```
row | C1   |  C2
------------------
1   |  X   |  Y
2   | null | CLEAR
```
The latest-at check for C1 protects row 1, and so if we GC row 2 then Y will come back into existence incorrectly.